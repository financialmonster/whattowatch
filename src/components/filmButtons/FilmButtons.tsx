import React, { FC, memo, useCallback, useEffect } from 'react';
import { push } from 'connected-react-router/immutable';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import { Map } from 'immutable';

import { useFetchFavorite } from 'hooks/useFetchFavorite';
import * as promoSelectors from 'domains/promo/promoSelectors';
import * as authSelectors from 'domains/auth/authSelectors';
import { isPromo } from 'utils';
import { favoritesActions } from 'domains/favorites/favoritesActions';
import { Notification } from 'components/notification/Notification';

type TFilmButtonsProps = {
    playBtnClickHandler: () => void;
    detailed?: boolean;
}

export const FilmButtons: FC<TFilmButtonsProps> = memo(({playBtnClickHandler, detailed}) => {
    const {id} = useParams();
    const dispatch = useDispatch();
    const promo = useSelector(promoSelectors.getPromo);
    const user = useSelector(authSelectors.getUser);

    useEffect(() => {
        dispatch(favoritesActions.resetFavoriteError());
    }, [dispatch]);
    
    const addReviewBtnClickHandler = useCallback(() => {
        dispatch(push(`${process.env.PUBLIC_URL}/film/${id}/review`));
    }, [dispatch, id]);

    let favorite: number | Map<string, any>;

    if(!id && isPromo(promo)) {
        favorite = promo;
    } else {
        favorite = +id;
    }

    const {myListBtnClickHandler, isFavorite, isFavoriteFetching, favoriteError} = useFetchFavorite(favorite);    

    return (
        <>
            <button className="btn btn--play movie-card__button" type="button" title="Play"
                onClick={playBtnClickHandler}
            >
                <svg viewBox="0 0 19 19" width="19" height="19">
                    <use xlinkHref="#play"></use>
                </svg>
                <span>Play</span>
            </button>
            <button className="btn btn--list movie-card__button" type="button" title="My list"
                onClick={myListBtnClickHandler}
            >
                <svg viewBox="0 0 19 20" width="19" height="20">
                    <use xlinkHref={(isFavorite && user) ? `#in-list` : `#add`}></use>
                </svg>
                <span>My list</span>
            </button>
            {(detailed) &&
                <a className="btn btn--add-review movie-card__button" title="Add review"
                    onClick={addReviewBtnClickHandler}
                >
                    Add review
                </a> 
            }
            {(isFavoriteFetching) &&
                <Notification>
                    Saving to the favorites list... Please, wait.
                </Notification>
            }
            {(favoriteError) &&
                <Notification>
                    {(favoriteError as Error).message}
                </Notification>
            }
        </>
    );
});
