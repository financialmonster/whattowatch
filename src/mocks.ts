import { Map, fromJS, List } from 'immutable';

export const createStoreMock = (mode: string = `common`) => {
    return {
        promo: Map({
            promo: Map(),
            isPromoFetching: true,
            promoError: null
        }),
        films: Map({
            films: fromJS([{genre: `Thriller`, id: 1}, {genre: `Comedy`, id: 2}]),
            isFilmsFetching: mode === `detailed`,
            filmsError: null,
            filter: `All genres`
        }),
        auth: Map({
            user: null,
            authError: null,
            isAuthFetching: false,
            authStatusError: null,
            isAuthStatusFetching: false
        }),
        reviews: Map({
            reviews: List(),
            isReviewsFetching: true,
            reviewsError: null
        })
    }
}
