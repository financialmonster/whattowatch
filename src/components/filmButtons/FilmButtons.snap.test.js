/* eslint-disable import/first */
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: () => ({id: 5})}
));

import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import { FilmButtons } from './FilmButtons';
import { createStoreMock } from 'mocks';

it(`FilmButtons: correctly rendered`, () => {
    const mockStore = configureStore();
    const store = mockStore(createStoreMock());

    const tree = renderer.create(
        <Provider store={store}>
            <FilmButtons />
        </Provider>        
    ).toJSON();

    expect(tree).toMatchSnapshot();
});
