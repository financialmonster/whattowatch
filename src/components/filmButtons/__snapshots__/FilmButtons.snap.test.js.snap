// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmButtons: correctly rendered 1`] = `
Array [
  <button
    className="btn btn--play movie-card__button"
    title="Play"
    type="button"
  >
    <svg
      height="19"
      viewBox="0 0 19 19"
      width="19"
    >
      <use
        xlinkHref="#play"
      />
    </svg>
    <span>
      Play
    </span>
  </button>,
  <button
    className="btn btn--list movie-card__button"
    onClick={[Function]}
    title="My list"
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 19 20"
      width="19"
    >
      <use
        xlinkHref="#add"
      />
    </svg>
    <span>
      My list
    </span>
  </button>,
]
`;
